nucleus/admin/util/src/main/java/com/sun/enterprise/admin/remote/reader/MultipartProprietaryReader.java

```diff
@@ -69,50 +72,54 @@ public ParamsWithPayload readFrom(final InputStream is, final String contentType
-                }
-                if (payload == null) {
-                    payload = new RestPayloadImpl.Inbound();
+        try (MIMEMessage mimeMessage = new MIMEMessage(is, boundary, new MIMEConfig())) {
+            // Parse
+            for (MIMEPart mimePart : mimeMessage.getAttachments()) {
+                String cd = getFirst(mimePart.getHeader("Content-Disposition"));
+                if (!StringUtils.ok(cd)) {
+                    cd = "file";
                 }
-                String ct = mimePart.getContentType();
-                if (!StringUtils.ok(ct) || ct.trim().startsWith("text/plain")) {
-                    payload.add(name, stream2String(mimePart.readOnce()), mimePart.getAllHeaders());
+                cd = cd.trim();
+                Properties cdParams = parseHeaderParams(cd);
+                // 3 types of content disposition
+                if (cd.startsWith("form-data")) {
+                    // COMMAND PARAMETER
+                    if (!StringUtils.ok(cdParams.getProperty("name"))) {
+                        throw new IOException("Form-data Content-Disposition does not contains name parameter.");
+                    }
+                    if (parameters == null) {
+                        parameters = new ParameterMap();
+                    }
+                    parameters.add(cdParams.getProperty("name"), stream2String(mimePart.readOnce()));
+                } else if (mimePart.getContentType() != null
+                    && mimePart.getContentType().startsWith("application/json")) {
+                    // ACTION REPORT
+                    actionReport = actionReportReader.readFrom(mimePart.readOnce(), "application/json");
                 } else {
-                    payload.add(name, mimePart.read(), ct, mimePart.getAllHeaders());
+                    // PAYLOAD
+                    String name = "noname";
+                    if (cdParams.containsKey("name")) {
+                        name = cdParams.getProperty("name");
+                        name = new String(name.getBytes("ISO8859-1"), "UTF-8");
+                    } else if (cdParams.containsKey("filename")) {
+                        name = cdParams.getProperty("filename");
+                    }
+                    if (payload == null) {
+                        payload = new RestPayloadImpl.Inbound();
+                    }
+                    String ct = mimePart.getContentType();
+                    if (!StringUtils.ok(ct) || ct.trim().startsWith("text/plain")) {
+                        payload.add(name, stream2String(mimePart.readOnce()), mimePart.getAllHeaders());
+                    } else {
+                        payload.add(name, mimePart.read(), ct, mimePart.getAllHeaders());
+                    }
                 }
             }
+            // Result
+            return new ParamsWithPayload(payload, parameters, actionReport);
+        } finally {
+            is.close();
```

Comment:
Is it safe to close the stream passed in as a method argument inside the method?

> public ParamsWithPayload readFrom(final InputStream is, final String contentType) throws IOException {

Reply:
The short answer is no and this  was reverted in the next commit. But it is even worse, this has to be rewritten completely, that is why I added this comment:
```
        // FIXME: Big refactoring required because of possible resource leaks:
        // MIMEMessage produces MIMEPart objects based on the is parameter.
        // All these objects are Closeable.
        // The final result of this method is the payload, based on mimePart object.
        // That is used later, and at least in one branch none of these objecs can be closed,
        // because the payload's input stream would be closed too.
```
The whole implementation is not "symmetric" for try-with, so close methods were originally "here and there" in the code. And still are and ie RemoteRestAdminCommand still can leak. I will revisit that in another iteration.